### PR TITLE
Use eslint for TS files

### LINKED ISSUE #7939: Update jupyterlab's lint setup
As discussed in a recent weekly dev meeting, the latest ESLint now supports a typescript plugin which would make tslint no longer necessary. the ipywidgets repo recently updated it's lint configs to the new version and can be used as a guide on updating the lab lint configs and version.

### PR DESCRIPTION
**Pull request by blink1073**
## References
Fixes #7939.  Switch to `@typescript-eslint`.  I used `npx tslint-to-eslint-config` to do the heavy lifting.  Adds linting for React (caught a few :bug:!).  Does not yet add jest linting because our mocha usage interferes. 
 

## Code changes
Combines tslint rules into eslint.  Updates lint-staged.

## User-facing changes
Better dev experience for maintainers.

## Backwards-incompatible changes
The following rules no longer apply (output from `tslint-to-eslint-config`):

```
tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "ban"
tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "jsdoc-format"
tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "one-line"
tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "prettier"
tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "whitespace"
```

### CODE REVIEW

**Comment by blink1073** [comment]
Closing this effort because typescript-eslint does not seem well suited to our package layout.  I am getting errors of the type:

```
/Users/stslve/workspace/jupyter/jlab-master/tests/webpack.config.js
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: tests/webpack.config.js.
The file must be included in at least one of the projects provided
```

I was also seeing OOM errors: https://github.com/typescript-eslint/typescript-eslint/issues/1192.
I think we should use `babel-eslint` instead, since we don't need our linter to also run a compiler.

**Comment by blink1073** [comment]
Note: we can't use `babel-eslint` because we rely on the `no-floating-promises` rule to catch bugs, and that requires a compiler.

**Comment by blink1073** [comment]
Pausing for now.  The linter is working locally.  It takes about 90s to parse all of our JS/TS source.
With the current set of rules, here's the output:

```
✖ 6924 problems (6635 errors, 289 warnings)
  6173 errors and 0 warnings potentially fixable with the `--fix` option.
```

The ones it will fix automatically are mostly `var/let` -> `const` and spacing/indent.

The ones it will not fix automatically are of the variety (example output line and number of occurrences):

```
 17:10  error  eval can be harmful                     no-eval
 2
   11:20  error  Interface name must be prefixed with "I"  @typescript-eslint/interface-name-prefix
 1
   364:11  error  Unexpected aliasing of 'this' to local variable      @typescript-eslint/no-this-alias
 3
  609:34  error  'widget.activeCellIndex' is assigned to itself      no-self-assign
 7
  894:78  error    Unexpected empty method 'render'                       @typescript-eslint/no-empty-function
 2
  474:11  error  Unreachable code                                     no-unreachable
 2
  1032:16  error  Unexpected constant condition                             no-constant-condition
 1
  735:9   error  Unnecessary try/catch wrapper                             no-useless-catch
 2
 
  44:3  error  Unexpected console statement  no-console
 63

  251:9  error  Promises must be handled appropriately  @typescript-eslint/no-floating-promises
 297
```

**Comment by blink1073** [comment]
cc @ajbozarth, this is ready for review.

**Comment by blink1073** [comment]
The actual lint application is in the third commit, you you may want to review the first two commits only :wink:

**Comment by blink1073** [comment]
I added React lint and it caught some real bugs!  I couldn't add jest lint yet because we have some tests that still aren't using jest and it confused the linter.

File: package.json [hunk 1]
```diff
@@ -51,17 +51,17 @@
...
     "get:dependency": "node buildutils/lib/get-dependency.js",
     "postinstall": "node scripts/ensure-buildutils.js",
     "integrity": "node scripts/ensure-buildutils.js && node buildutils/lib/ensure-repo.js",
     "lighthouse": "lighthouse http://localhost:8888/ --chrome-flags='--headless' --emulated-form-factor=none --throttling-method=provided --only-categories performance",
     "lighthouse:compare": "node testutils/lib/compare-lighthouse.js",
     "lighthouse:throttling:start": "comcast --latency=40 --target-bw=30000 --packet-loss=0.2%",
     "lighthouse:throttling:stop": "comcast --stop",
-    "lint": "jlpm && jlpm run prettier && jlpm run eslint && jlpm run tslint",
-    "lint:check": "jlpm run prettier:check && jlpm run eslint:check && jlpm run tslint:check",
+    "lint": "jlpm && jlpm run eslint && jlpm run prettier",
```
Comment (ajbozarth) [inline package.json:63]: why swap the run order of prettier and eslint, iirc running prettier first is the standard

Comment (blink1073) [reply package.json:63]: eslint was undoing some of the things prettier was doing with a previous set of rules.  I'll revert this change.  Cheers!

**Comment by blink1073** [comment]
Stumbled?  Ha, more like staggered my way through it :smile:

**Pull request merged**